packages/duckdb-wasm/src/utils/s3_helper.ts

```diff
@@ -0,0 +1,221 @@
+        "/aws4_request\n" + canonical_request_hash_str;
+
+    // ts-ignore's because library can accept array buffer as key, but TS arg is incorrect
+    const sign_key = "AWS4" + params.secret_access_key;
+    const k_date = sha256.hmac.arrayBuffer(sign_key, params.date_now);
+
+    // Note, js-sha256 has a bug in the TS interface that only supports strings as keys, while we need a bytearray
+    // as key. PR is open but unmerged: https://github.com/emn178/js-sha256/pull/25
+    // eslint-disable-next-line
+    // @ts-ignore
+    const k_region = sha256.hmac.arrayBuffer(k_date, params.region);
+    // eslint-disable-next-line
+    // @ts-ignore
+    const k_service = sha256.hmac.arrayBuffer(k_region, params.service,);
+    // eslint-disable-next-line
+    // @ts-ignore
+    const signing_key = sha256.hmac.arrayBuffer(k_service, "aws4_request");
+    // eslint-disable-next-line
+    // @ts-ignore
+    const signature = sha256.hmac(signing_key, string_to_sign);
+
+    res.set("Authorization", "AWS4-HMAC-SHA256 Credential=" + params.access_key_id + "/" + params.date_now + "/" + params.region + "/" +
+        params.service + "/aws4_request, SignedHeaders=" + signed_headers +
+        ", Signature=" + signature);
+
+    return res;
+}
+
+const createS3HeadersFromS3Config = function (config : S3Config | undefined, url : string, method : string) : Map<string, string> {
+    const params = getS3Params(config, url, method);
+    return createS3Headers(params);
+}
+
+export function addS3Headers(xhr: XMLHttpRequest, config : S3Config | undefined, url : string, method: string) {
+    if (config?.accessKeyId || config?.sessionToken) {
+        const headers = createS3HeadersFromS3Config(config, url, method);
+        headers.forEach((value: string, header: string) => {
+            xhr.setRequestHeader(header, value);
+        });
+    }
+}
+
+export function parseS3Url (url: string) : {bucket : string, path : string} {
+    if (url.indexOf("s3://") != 0) {
+        throw new Error("URL needs to start with s3://");
+    }
+    const slash_pos = url.indexOf('/', 5);
+
+    if (slash_pos == -1) {
+        throw new Error("URL needs to contain a '/' after the host");
+    }
+
+    const bucket = url.substring(5, slash_pos);
+    if (!bucket) {
+        throw new Error("URL needs to contain a bucket name");
+    }
+    const path = url.substring(slash_pos);
+    if (!path) {
+        throw new Error("URL needs to contain key");
+    }
+
+    return {bucket: bucket, path: path}
+}
+
+export function getHTTPUrl(config : S3Config | undefined, url : string) : string {
+    const parsedUrl = parseS3Url(url);
+    const host = parsedUrl.bucket + "." + (config?.endpoint ? config?.endpoint : "s3.amazonaws.com");
+    const http_host = "https://" + host;
+    return http_host + parsedUrl.path;
+}
+
+// this computes the signature from https://czak.pl/2015/09/15/s3-rest-api-with-curl.html
+export function verifyS3Helper() {
```

Comment:
Can we move this to the test file? Otherwise this will be bundled within the library.